### PR TITLE
searching through deep arrays

### LINKED ISSUE #32: Index fields from nested array of objects?
Given a document like:

``` json
{
  "some_field": "foo",
  "meta": [
    {
       "nested": "First meta"
    },
    {
      "nested": "Second meta"
     }]
}
```

How can I index the "First meta" and "Second meta" values? I tried `meta.nested` and `meta[].nested` but they do not seem to work.

### PR DESCRIPTION
**Pull request by spicemix**
nnarhinen's implementation (#33) with nolanlawson's requested changes. Closes #32.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -39,7 +39,12 @@ function getText(fieldBoost, doc) {
   } else { // "Enhance."
     text = doc;
     for (var i = 0, len = fieldBoost.deepField.length; i < len; i++) {
-      text = text && text[fieldBoost.deepField[i]];
+      if (Array.isArray(text)) {
+        text = text.map(handleNestedObjectArrayItem(fieldBoost, fieldBoost.deepField.slice(i)));
+        break;
```
Comment (nolanlawson) [inline lib/index.js]: this `break` is not necessary

File: lib/index.js [hunk 2]
```diff
@@ -52,6 +57,14 @@ function getText(fieldBoost, doc) {
   return text;
 }
 
+function handleNestedObjectArrayItem(fieldBoost, deepField) {
+  return function (one) {
+    return getText(utils.extend({}, fieldBoost, {
+      deepField: deepField
+    }), one);
+  };
```
Comment (nolanlawson) [inline lib/index.js]: is it necessary to make this a function that returns a function? typically better for perf to just declare one function once

File: test/test.js [hunk 3]
```diff
@@ -564,6 +565,18 @@ function tests(dbName, dbType) {
         ids.should.deep.equal(['2']);
       });
     });
+    it('allows searching from an array of nested objects', function () {
+      return db.bulkDocs({docs: docs9}).then(function () {
+        var opts = {
+          fields: ['nested.array.aField'],
```
Comment (nolanlawson) [inline test/test.js]: I was a bit confused at first that this was conflating deep fields and array search, but I understand now that both features are supported independently. :)

File: lib/index.js (hunk 1 above)
Comment (spicemix) [inline lib/index.js]: Agreed, you can delete that. Unless you want me to update the PR for it. Tests all run fine with it deleted here.

File: lib/index.js (hunk 2 above)
Comment (spicemix) [inline lib/index.js]: He wrote this in a functional style where the function returns a customized function to feed into the map call. That way it maps back to your getText() function and any logic it has in it. It's not my style of coding and I don't have much insight on it but it works...

**Comment by caleuanhopkins** [comment]
+1

**Comment by rafiyajaved** [comment]
Hello! Also looking to deep search arrays here and came across this. Just wondering if this branch is going to get merged after all or if it still required changes?

**Comment by nolanlawson** [comment]
Sorry, forgot about this PR! 😅  I'll merge and publish today.

**Comment by nolanlawson** [comment]
published in v1.3.0

**Pull request merged**